> Approving, but I've not been able to test thoroughly. If i run it locally, the errors don't seem to be caught, and if I try to run on an instance, I keep running into issues with aws cli atm (unrelated to this PR)
> 
> ```
> $ aws configure
> Traceback (most recent call last):
>   File "/usr/bin/aws", line 19, in <module>
>     import awscli.clidriver
>   File "/usr/lib/python3/dist-packages/awscli/clidriver.py", line 36, in <module>
>     from awscli.help import ProviderHelpCommand
>   File "/usr/lib/python3/dist-packages/awscli/help.py", line 23, in <module>
>     from botocore.docs.bcdoc import docevents
> ImportError: cannot import name 'docevents' from 'botocore.docs.bcdoc' (/home/ubuntu/.local/lib/python3.8/site-packages/botocore/docs/bcdoc/__init__.py)
> ```
> 
> Generally though I think you should remove the reliance on boto for getting the instance name. I think it would be better for this to be small and simple, without having to worry about aws credentials, or boto issues like I am experiencing!

I agree let’s simplify. We can always reintroduce if needed. I will record the issues you experienced regardless

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -11,4 +11,6 @@ jobs:
       - name: "Install package"
         run: pip install .[test]
       - name: "Run tests"
+        env:
+          BELL_MODE: "LOCAL"
         run: pytest

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -1,0 +1,14 @@
+name: "Run tests"
+on: ["push"]
+jobs:
+  Run-tests:
+    runs-on: ["ubuntu-latest"]
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: "3.8"
+      - name: "Install package"
+        run: pip install .[test]
+      - name: "Run tests"
+        run: pytest

--- a/README.md
+++ b/README.md
@@ -7,24 +7,38 @@ A tool to send you alerts from your AWS instances
 Install
 
 ```
-pip install git+https://github.com/MantisAI/bell.git
+pip install git+https://github.com/MantisAI/bell.git@main
 ```
 
-Use in an EC2 instance
+Use bell 🛎  to get alerts that your AWS instance is on
 
 ```
-bell SLACK_WEBHOOK
+bell --webhook-url SLACK_WEBHOOK_URL
 ```
+
+Use bell 🛎  to get alerts about when a command starts, fails or finishes
+
+```
+export SLACK_WEBHOOK_URL=xxx
+
+bell python -c "print('hello')"
+```
+
+Note that you can pass the slack webhook URL either as a command line parameter
+`--webhook-url` or as an environment variable `SLACK_WEBHOOK_URL`
 
 You can read how to setup a slack webbook [here](https://api.slack.com/incoming-webhooks)
 
+
 # Cron
 
-This tool is intended to be used with a scheduler like cron. Here is an example cron
+This tool works well with a scheduler like cron. Here is an example cron
 command which sends an alert every 2 hours
 
 ```
-0 */2 * * * bell WEBHOOK_URL
+SLACK_WEBHOOK_URL=xxx
+
+0 */2 * * * bell
 ```
 
 Note that cron's environment is different to your user or root environment. See troubleshooting
@@ -33,13 +47,12 @@ if you have problems setting up bell to work with cron.
 # CLI
 
 ```
-Usage: bell [OPTIONS] WEBHOOK_URL
+usage: bell [-h] [--webhook-url WEBHOOK_URL]
 
-Arguments:
-  WEBHOOK_URL  [required]
-
-Options:
-  --help                          Show this message and exit.
+optional arguments:
+  -h, --help            show this help message and exit
+  --webhook-url WEBHOOK_URL
+                        incoming slack webhook url
 ```
 
 # Troubleshooting

--- a/bell.py
+++ b/bell.py
@@ -7,7 +7,8 @@ import requests
 import boto3
 
 
-BELL_MODE = os.environ["BELL_MODE"]
+BELL_MODE = os.environ.get("BELL_MODE")
+
 
 def get_status_message():
     if BELL_MODE == "LOCAL":
@@ -37,9 +38,7 @@ def send_slack_message(message):
     if BELL_MODE == "LOCAL":
         return
 
-    response = requests.post(
-        webhook_url, json={"text": message}
-    )
+    response = requests.post(webhook_url, json={"text": message})
     logger.debug(f"slack response code: {response.status_code}")
 
 

--- a/bell.py
+++ b/bell.py
@@ -1,14 +1,18 @@
+import subprocess
+import argparse
+import os
+
 from loguru import logger
 import requests
-import typer
 import boto3
 
 
-app = typer.Typer()
+BELL_MODE = os.environ["BELL_MODE"]
 
+def get_status_message():
+    if BELL_MODE == "LOCAL":
+        return ":bellhop_bell: local mode"
 
-@app.command()
-def bell(webhook_url):
     response = requests.get("http://169.254.169.254/latest/meta-data/instance-id")
     instance_id = response.text
 
@@ -24,12 +28,48 @@ def bell(webhook_url):
     response = ec2.describe_instance_status(InstanceIds=[instance_id])
     status = response["InstanceStatuses"][0]["InstanceState"]["Name"]
     logger.debug(f"ec2 instance status: {status}")
+    return f":bellhop_bell: {ec2_instance_name} is {status}"
+
+
+def send_slack_message(message):
+    logger.debug(f"slack message: {message}")
+
+    if BELL_MODE == "LOCAL":
+        return
 
     response = requests.post(
-        webhook_url, json={"text": f":bellhop_bell: {ec2_instance_name} is {status}"}
+        webhook_url, json={"text": message}
     )
     logger.debug(f"slack response code: {response.status_code}")
 
 
+def bell(webhook_url, *command_args):
+    if command_args:
+        logger.debug(f"command: {command_args}")
+        try:
+            send_slack_message("command started")
+            subprocess.run(command_args, check=True)
+        except subprocess.CalledProcessError:
+            logger.error("command failed")
+            send_slack_message("command failed")
+            raise
+
+        send_slack_message("command finished")
+    else:
+        instance_status = get_status_message()
+        send_slack_message(instance_status)
+
+
+def cli():
+    argument_parser = argparse.ArgumentParser()
+    argument_parser.add_argument("--webhook-url", help="incoming slack webhook url")
+
+    args, command_args = argument_parser.parse_known_args()
+    logger.debug(f"args: {args}")
+    logger.debug(f"command_args: {command_args}")
+
+    bell(args.webhook_url, *command_args)
+
+
 if __name__ == "__main__":
-    app()
+    cli()

--- a/setup.cfg
+++ b/setup.cfg
@@ -18,10 +18,13 @@ classifiers =
 install_requires =
     boto3
     requests
-    typer
     loguru
 python_requires = >=3.6
 
 [options.entry_points]
 console_scripts =
-    bell = bell:app
+    bell = bell:cli
+
+[options.extras_require]
+test =
+    pytest

--- a/tests/test_bell.py
+++ b/tests/test_bell.py
@@ -1,0 +1,18 @@
+import subprocess
+
+import pytest
+
+from bell import bell
+
+
+def test_bell():
+    bell("slack webhook url")
+
+def test_bell_command():
+    command_args = "python -c \"print('hello')\"".split()
+    bell("slack webhook url", *command_args)
+
+def test_bell_bad_command():
+    command_args = """ python -c "print('hello' """.split()
+    with pytest.raises(subprocess.CalledProcessError):
+        bell("slack webhook url", *command_args)

--- a/tests/test_bell.py
+++ b/tests/test_bell.py
@@ -8,9 +8,11 @@ from bell import bell
 def test_bell():
     bell("slack webhook url")
 
+
 def test_bell_command():
     command_args = "python -c \"print('hello')\"".split()
     bell("slack webhook url", *command_args)
+
 
 def test_bell_bad_command():
     command_args = """ python -c "print('hello' """.split()

